<a href="https://colab.research.google.com/github/satishgunjal/Real-Estate-Price-Prediction-Project/blob/master/House_Price_Prediction_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib
matplotlib.rcParams['figure.figsize'] = (20,10) # width, height in inches
from google.colab import files
import io

In [ ]:
# for local notebook
#df1 = pd.read_csv('Bengaluru_House_Data.csv')
#df1.head()

In [4]:
#For google colab
uploaded = files.upload()

Saving Bengaluru_House_Data.csv to Bengaluru_House_Data.csv


In [5]:
df1 = pd.read_csv(io.StringIO(uploaded['Bengaluru_House_Data.csv'].decode('utf-8')))
df1.head()

area_type   availability  ... balcony   price
0  Super built-up  Area         19-Dec  ...     1.0   39.07
1            Plot  Area  Ready To Move  ...     3.0  120.00
2        Built-up  Area  Ready To Move  ...     3.0   62.00
3  Super built-up  Area  Ready To Move  ...     1.0   95.00
4  Super built-up  Area  Ready To Move  ...     1.0   51.00

[5 rows x 9 columns]

In [6]:
# Get the no of rows and columns
df1.shape

(13320, 9)

In [7]:
#Get all the column names
df1.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

In [8]:
#Lets check the unique values 'area_type' column
df1.area_type.unique()

array(['Super built-up  Area', 'Plot  Area', 'Built-up  Area',
       'Carpet  Area'], dtype=object)

In [9]:
#Let get the count of trianing examples for each area type
df1.area_type.value_counts()

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64

In [10]:
# Note everytime we make change in dataset we store it in new dataframe
df2 = df1.drop(['area_type', 'availability', 'society', 'balcony'],axis='columns')

print('Rows and columns are = ', df2.shape)
df2.head()

Rows and columns are =  (13320, 5)


location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [11]:
# Get the sum of all na values from dataset
df2.isna().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

Since null values as comapre to total training examples(13320) is verry less we can safly drop those examples

In [12]:
df3 = df2.dropna()
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [13]:
# Since all oor training examples containing null values are dropped lets check the shape of the dataset again
df3.shape

(13246, 5)

In [14]:
df3['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

In [15]:
df4 = df3.copy()

# Using lambda function we can get the BHK numeric value
df4['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))
#df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))
df4.bhk.unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18])

From above data we can see that there are home with upto 43 BHK's in Bangalore.. must be apolitician :)

In [16]:
#Get the training examples with home size more than 20 BHK
df4[df4.bhk >20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

Note above 43 BHK home area is only 2400 sqrft only. We will remove this data error later. First lets clean the 'total_sqft' column

Now lets check the unique values in 'total_sqft' column

In [17]:
df4.total_sqft.unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

Note above, there are few records with range of the area like '1133 - 1384'.
Lets write a function to identify such values

In [ ]:
def is_float(x):
  try:
    float(x)
  except:
    return False

  return True

In [19]:
# Test the function
print('is this (123) float value = %s' % (is_float(123)))
print('is this (1133 - 1384) float value = %s' % (is_float('1133 - 1384')))

is this (123) float value = True
is this (1133 - 1384) float value = False


In [20]:
#Lets apply this function to 'total_sqft' column

#Showing training examples where 'total_sqft' vale is not float
df4[~df4['total_sqft'].apply(is_float)].head(10) 

location       size      total_sqft  bath    price  bhk
30            Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122              Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137  8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165            Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188            KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410             Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549         Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648             Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661           Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672        Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4

* Since most the value are range of sqft, we can write afunction to get the average value from a range. 
* There are few values like '34.46Sq. Meter' and '4125Perch' we can also try and convert those values into sqft but for now I amgoing to ignore them

In [ ]:
def convert_range_to_sqft(x):
  try:
    tokens = x.split('-')

    if len(tokens) == 2:
      return (float(tokens[0]) + float(tokens[1]))/2
    else:
      return float(x)
  except:
    return None

In [22]:
#Lets test the convert_range_to_sqft()
print('Return value for i/p 12345 = %s' % (convert_range_to_sqft('12345')))
print('Return value for i/p 1133 - 1384 = %s' % (convert_range_to_sqft('1133 - 1384')))
print('Return value for i/p 34.46Sq. Meter = %s' % (convert_range_to_sqft('34.46Sq. Meter')))

Return value for i/p 12345 = 12345.0
Return value for i/p 1133 - 1384 = 1258.5
Return value for i/p 34.46Sq. Meter = None


In [23]:
# Lets apply this function for total_sqft column
df5 = df4.copy()

df5.total_sqft = df4.total_sqft.apply(convert_range_to_sqft)
df5.head()

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

In [24]:
# Since our converion function will return null for values like 34.46Sq. Meter. Lets check for any null values in it
df5.total_sqft.isnull().sum()

46

In [25]:
# Lets dro the null training sets from total_sqft
df6 = df5.dropna()
df6.total_sqft.isnull().sum()

# OR
#We can also select the not null training set using below filter
#df6 = df5[df5.total_sqft.notnull()]

0

In [26]:
# Lets cross check the values of 'total_sqft'
print('total_sqft value for 30th training set in df4 = %s' % (df4.total_sqft[30]))
print('total_sqft value for 30th training set in df6 = %s' % (df6.total_sqft[30]))

total_sqft value for 30th training set in df4 = 2100 - 2850
total_sqft value for 30th training set in df6 = 2475.0


## Feature Engineering
* 'price' column containes the price of house in lacka ( 1 lakh = 100000)
* Price per square fit is important parameter in house prices.
* So we can create new column by the name 'price_per_sqft' and add price per sqft in it. formula = (price * 100000)/total_sqft

In [27]:
df7 = df6.copy()

df7['price_per_sqft'] = (df6['price'] * 100000)/df6['total_sqft']
df7.head()

location       size  total_sqft  ...   price  bhk  price_per_sqft
0  Electronic City Phase II      2 BHK      1056.0  ...   39.07    2     3699.810606
1          Chikka Tirupathi  4 Bedroom      2600.0  ...  120.00    4     4615.384615
2               Uttarahalli      3 BHK      1440.0  ...   62.00    3     4305.555556
3        Lingadheeranahalli      3 BHK      1521.0  ...   95.00    3     6245.890861
4                  Kothanur      2 BHK      1200.0  ...   51.00    2     4250.000000

[5 rows x 7 columns]

In [28]:
df7_stats = df7['price_per_sqft'].describe()
df7_stats

count    1.320000e+04
mean     7.920759e+03
std      1.067272e+05
min      2.678298e+02
25%      4.267701e+03
50%      5.438331e+03
75%      7.317073e+03
max      1.200000e+07
Name: price_per_sqft, dtype: float64

In [29]:
#Trim the location values
df7.location = df7.location.apply(lambda x: x.strip())
df7.head()

location       size  total_sqft  ...   price  bhk  price_per_sqft
0  Electronic City Phase II      2 BHK      1056.0  ...   39.07    2     3699.810606
1          Chikka Tirupathi  4 Bedroom      2600.0  ...  120.00    4     4615.384615
2               Uttarahalli      3 BHK      1440.0  ...   62.00    3     4305.555556
3        Lingadheeranahalli      3 BHK      1521.0  ...   95.00    3     6245.890861
4                  Kothanur      2 BHK      1200.0  ...   51.00    2     4250.000000

[5 rows x 7 columns]

In [30]:
#Lets get the count of each location
location_stats = df7.location.value_counts(ascending=False)
location_stats

Whitefield                                     533
Sarjapur  Road                                 392
Electronic City                                304
Kanakpura Road                                 264
Thanisandra                                    235
                                              ... 
Somanna Garden                                   1
Howthinarayanappa Garden                         1
Bharat Nagar                                     1
Annapoorneshwari Layout, JP nagar 7th phase      1
Kalkere Channasandra                             1
Name: location, Length: 1287, dtype: int64

In [31]:
#Total number unique location categories are
len(location_stats)

1287

We are going assign a category 'other' for every location where total datapoints are less than 10

In [32]:
#Get total number of categories where data points are less than 10
print('Total no of locations where data points are more than 10 = %s' % (len(location_stats[location_stats > 10])))
print('Total no of locations where data points are less than 10 = %s' % (len(location_stats[location_stats <= 10])))

Total no of locations where data points are more than 10 = 240
Total no of locations where data points are less than 10 = 1047


Any location having less than 10 data points should be tagged as "other" location. This way number of categories can be reduced by huge amount. Later on when we do one hot encoding, it will help us with having fewer dummy columns

In [33]:
location_stats_less_than_10 = location_stats[location_stats <= 10]
location_stats_less_than_10

Gunjur Palya                                   10
Nagappa Reddy Layout                           10
Sector 1 HSR Layout                            10
Ganga Nagar                                    10
Naganathapura                                  10
                                               ..
Somanna Garden                                  1
Howthinarayanappa Garden                        1
Bharat Nagar                                    1
Annapoorneshwari Layout, JP nagar 7th phase     1
Kalkere Channasandra                            1
Name: location, Length: 1047, dtype: int64

In [34]:
#Using lambda function assign the 'other' type to every element in 'location_stats_less_than_10'
df8 = df7.copy()

df8.location = df7.location.apply(lambda x: 'other' if x in location_stats_less_than_10 else x )
len(df8.location.unique())

241

Since 1047 location with less than 10 data points are converted to one category 'other'
Total no of unique location categories are = 240 +1  = 241

In [35]:
df8.head(10)

location       size  total_sqft  ...   price  bhk  price_per_sqft
0  Electronic City Phase II      2 BHK      1056.0  ...   39.07    2     3699.810606
1          Chikka Tirupathi  4 Bedroom      2600.0  ...  120.00    4     4615.384615
2               Uttarahalli      3 BHK      1440.0  ...   62.00    3     4305.555556
3        Lingadheeranahalli      3 BHK      1521.0  ...   95.00    3     6245.890861
4                  Kothanur      2 BHK      1200.0  ...   51.00    2     4250.000000
5                Whitefield      2 BHK      1170.0  ...   38.00    2     3247.863248
6          Old Airport Road      4 BHK      2732.0  ...  204.00    4     7467.057101
7              Rajaji Nagar      4 BHK      3300.0  ...  600.00    4    18181.818182
8              Marathahalli      3 BHK      1310.0  ...   63.25    3     4828.244275
9                     other  6 Bedroom      1020.0  ...  370.00    6    36274.509804

[10 rows x 7 columns]

In [36]:
# Lets visualize the data where square fit per bedroom is less than 300
df8[(df8.total_sqft / df8.bhk) < 300]

location       size  total_sqft  ...  price  bhk  price_per_sqft
9                    other  6 Bedroom      1020.0  ...  370.0    6    36274.509804
45              HSR Layout  8 Bedroom       600.0  ...  200.0    8    33333.333333
58           Murugeshpalya  6 Bedroom      1407.0  ...  150.0    6    10660.980810
68     Devarachikkanahalli  8 Bedroom      1350.0  ...   85.0    8     6296.296296
70                   other  3 Bedroom       500.0  ...  100.0    3    20000.000000
...                    ...        ...         ...  ...    ...  ...             ...
13277                other  7 Bedroom      1400.0  ...  218.0    7    15571.428571
13279                other  6 Bedroom      1200.0  ...  130.0    6    10833.333333
13281      Margondanahalli  5 Bedroom      1375.0  ...  125.0    5     9090.909091
13303       Vidyaranyapura  5 Bedroom       774.0  ...   70.0    5     9043.927649
13311     Ramamurthy Nagar  7 Bedroom      1500.0  ...  250.0    7    16666.666667

[744 rows x 7 columns]

In [37]:
#Lets check current dataset shape before removing outliers
df8.shape

(13200, 7)

In [38]:
df9 = df8[~((df8.total_sqft / df8.bhk) < 300)]
df9.shape

(12456, 7)

In [39]:
# Get basic stats of column 'price_per_sqft'
df9.price_per_sqft.describe()

count     12456.000000
mean       6308.502826
std        4168.127339
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_sqft, dtype: float64

Note: Its important to understand that price of every house is location specific. We are going to remove outliers using 'price_per_sqft' for each location

In [41]:
#Lets check current dataset shape before removing outliers
df9.shape

(12456, 7)

In [42]:
# Function to remove outliers using pps(price per sqft)
def remove_pps_outliers(df):
    df_out = pd.DataFrame()
    for key, subdf in df.groupby('location'):
        mean = np.mean(subdf.price_per_sqft)
        std = np.std(subdf.price_per_sqft)
        reduced_df = subdf[(subdf.price_per_sqft>(mean-std)) & (subdf.price_per_sqft<=(mean + std))] # 1 Sigma value i.e 68% of data
        df_out = pd.concat([df_out,reduced_df],ignore_index=True) # Storing data in 'df_out' dataframe
    return df_out

df10 = remove_pps_outliers(df9)
df10.shape

(10242, 7)

In [46]:
def remove_bhk_outliers(df):
    exclude_indices = np.array([])
    for location, location_df in df.groupby('location'):
        bhk_stats = {}
        for bhk, bhk_df in location_df.groupby('bhk'):
            bhk_stats[bhk] = {
                'mean': np.mean(bhk_df.price_per_sqft),
                'std': np.std(bhk_df.price_per_sqft),
                'count': bhk_df.shape[0]
            }
        for bhk, bhk_df in location_df.groupby('bhk'):
            stats = bhk_stats.get(bhk-1)
            if stats and stats['count']>5:
                exclude_indices = np.append(exclude_indices, bhk_df[bhk_df.price_per_sqft<(stats['mean'])].index.values)
    return df.drop(exclude_indices,axis='index')

df11 = remove_bhk_outliers(df10)
df11.shape

(7317, 7)

Now you can campre the scatter plots for location(Hebbal and Rajaji Nagar) for before and after outlier removal

In [50]:
#Get unique bath from dataset
df11.bath.unique()

array([ 4.,  3.,  2.,  5.,  8.,  1.,  6.,  7.,  9., 12., 16., 13.])

In [51]:
#Get the training examples where no of bath are more than (no of BHK +2)
df11[df11.bath > df11.bhk + 2]

location       size  total_sqft  bath   price  bhk  price_per_sqft
1626  Chikkabanavar  4 Bedroom      2460.0   7.0    80.0    4     3252.032520
5238     Nagasandra  4 Bedroom      7000.0   8.0   450.0    4     6428.571429
6711    Thanisandra      3 BHK      1806.0   6.0   116.0    3     6423.034330
8408          other      6 BHK     11338.0   9.0  1000.0    6     8819.897689

We can remove above outliers from the datset

In [52]:
#Lets check current dataset shape before removing outliers
df11.shape

(7317, 7)

In [53]:
# Remove the outliers with more than (no of BHK + 2) bathrooms
df12 = df11[df11.bath < (df11.bhk + 2)]
df12.shape

(7239, 7)

This concludes our data cleaning, lets drop unnecessary columns
* Since we have 'bhk' feature lets drop 'size'
* We have crerated 'price_per_sqft' for outlier detection and removal purpose, so we can also drop it. 

In [54]:
df13 = df11.drop(['size', 'price_per_sqft'], axis='columns')
df13.head()

location  total_sqft  bath  price  bhk
0  1st Block Jayanagar      2850.0   4.0  428.0    4
1  1st Block Jayanagar      1630.0   3.0  194.0    3
2  1st Block Jayanagar      1875.0   2.0  235.0    3
3  1st Block Jayanagar      1200.0   2.0  130.0    3
4  1st Block Jayanagar      1235.0   2.0  148.0    2

In [55]:
dummies = pd.get_dummies(df13.location)
dummies.head()

1st Block Jayanagar  1st Phase JP Nagar  ...  Yeshwanthpur  other
0                    1                   0  ...             0      0
1                    1                   0  ...             0      0
2                    1                   0  ...             0      0
3                    1                   0  ...             0      0
4                    1                   0  ...             0      0

[5 rows x 241 columns]

In [56]:
#To avoid dummy variable trap problem lets delete the one of the dummy variable column
dummies = dummies.drop(['other'],axis='columns')
dummies.head()

1st Block Jayanagar  1st Phase JP Nagar  ...  Yelenahalli  Yeshwanthpur
0                    1                   0  ...            0             0
1                    1                   0  ...            0             0
2                    1                   0  ...            0             0
3                    1                   0  ...            0             0
4                    1                   0  ...            0             0

[5 rows x 240 columns]

In [57]:
#Now lets add dummies dataframe to original dataframe
df14 = pd.concat([df13,dummies],axis='columns')
df14.head()

location  total_sqft  ...  Yelenahalli  Yeshwanthpur
0  1st Block Jayanagar      2850.0  ...            0             0
1  1st Block Jayanagar      1630.0  ...            0             0
2  1st Block Jayanagar      1875.0  ...            0             0
3  1st Block Jayanagar      1200.0  ...            0             0
4  1st Block Jayanagar      1235.0  ...            0             0

[5 rows x 245 columns]

In [58]:
#Lets delete the location feature
df15 = df14.drop(['location'],axis='columns')
df15.head()m

total_sqft  bath  price  ...  Yelahanka New Town  Yelenahalli  Yeshwanthpur
0      2850.0   4.0  428.0  ...                   0            0             0
1      1630.0   3.0  194.0  ...                   0            0             0
2      1875.0   2.0  235.0  ...                   0            0             0
3      1200.0   2.0  130.0  ...                   0            0             0
4      1235.0   2.0  148.0  ...                   0            0             0

[5 rows x 244 columns]

#Step#5: Build Machine Learning Model


In [60]:
#Final shape of our dataset is
df15.shape

(7317, 244)

Now leats create X(independent variable/features) and y(dependent variables/target)

In [63]:
X = df15.drop(['price'],axis='columns')
X.head()

total_sqft  bath  bhk  ...  Yelahanka New Town  Yelenahalli  Yeshwanthpur
0      2850.0   4.0    4  ...                   0            0             0
1      1630.0   3.0    3  ...                   0            0             0
2      1875.0   2.0    3  ...                   0            0             0
3      1200.0   2.0    3  ...                   0            0             0
4      1235.0   2.0    2  ...                   0            0             0

[5 rows x 243 columns]

In [62]:
y = df15.price
y.head()

0    428.0
1    194.0
2    235.0
3    130.0
4    148.0
Name: price, dtype: float64

In [68]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=10)
print('X_train shape = ',X_train.shape)
print('X_test shape = ',X_test.shape)
print('y_train shape = ',y_train.shape)
print('y_test shape = ',y_test.shape)

X_train shape =  (5853, 243)
X_test shape =  (1464, 243)
y_train shape =  (5853,)
y_test shape =  (1464,)


In [85]:
from sklearn.linear_model import LinearRegression

lr_clf = LinearRegression()
lr_clf.fit(X_train, y_train)
lr_clf.score(X_test, y_test)

0.8175889965469311

In [70]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit

# ShuffleSplit is used to randomize the each fold
cv = ShuffleSplit(n_splits=5, test_size=0.2, random_state=0)
cross_val_score(LinearRegression(), X, y, cv = cv)

array([0.83458131, 0.8720703 , 0.77477801, 0.86043362, 0.87857309])

In [77]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Lasso
from sklearn.tree import DecisionTreeRegressor

from sklearn.linear_model import Ridge
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import ElasticNet

def find_best_model_using_gridsearchcv(X,y):
  algos ={
      'linear_regression':{
          'model':LinearRegression(),
          'params': {
              'normalize':[True,False]
          }
      },
        'lasso': {
            'model': Lasso(),
            'params': {
                'alpha': [1,2],
                'selection': ['random', 'cyclic']
            }
        },
        'decision_tree': {
            'model': DecisionTreeRegressor(),
            'params': {
                'criterion' : ['mse','friedman_mse'],
                'splitter': ['best','random']
            }
        }
  }

  scores= []
  cv = ShuffleSplit(n_splits=5, test_size=0.2, random_state=0)

  for algo_name, config in algos.items():
      gs =  GridSearchCV(config['model'], config['params'], cv=cv, return_train_score=False)
      gs.fit(X,y)
      scores.append({
          'model': algo_name,
          'best_score': gs.best_score_,
          'best_params': gs.best_params_
      })

  return pd.DataFrame(scores,columns=['model','best_score','best_params'])

find_best_model_using_gridsearchcv(X,y)

model  ...                                        best_params
0  linear_regression  ...                               {'normalize': False}
1              lasso  ...                {'alpha': 1, 'selection': 'cyclic'}
2      decision_tree  ...  {'criterion': 'friedman_mse', 'splitter': 'ran...

[3 rows x 3 columns]

In [ ]:
def predict_price(location, sqft, bath, bhk):    
    loc_index = np.where(X.columns==location)[0][0]

    x = np.zeros(len(X.columns))
    x[0] = sqft
    x[1] = bath
    x[2] = bhk
    if loc_index >= 0:
        x[loc_index] = 1

    return lr_clf.predict([x])[0]

In [86]:
predict_price('1st Phase JP Nagar',1000, 2, 2)

83.16761667762826

In [87]:
predict_price('1st Phase JP Nagar',1000, 3, 3)

83.71323640897404

In [88]:
predict_price('Indira Nagar',1000, 2, 2)

158.07409801131223

In [89]:
predict_price('Indira Nagar',1000, 3, 3)

158.61971774265803

Step#7: Export the model to Pickle file

In [ ]:
import pickle

with open('Real_Estate_Price_Prediction_Project.pickle','wb') as f:
    pickle.dump(lr_clf,f)

In [94]:
#Since we are using the Google colab, pickle will will be saved at current directory of Google cloud machine
import os

os.listdir('.')

['.config',
 'Bengaluru_House_Data.csv',
 'Real_Estate_Price_Prediction_Project.pickle',
 'sample_data']

In [ ]:
#Lets download it
from google.colab import files

files.download('Real_Estate_Price_Prediction_Project.pickle')

Step#8: Export any other important info
* since weare using One Hot Encoding for location column we need the final list of all the columns in our feature set

In [ ]:
import json
columns = {
    'data_columns' : [col.lower() for col in X.columns]
}
with open("columns.json","w") as f:
    f.write(json.dumps(columns))

In [97]:
os.listdir('.')

['.config',
 'Bengaluru_House_Data.csv',
 'Real_Estate_Price_Prediction_Project.pickle',
 'columns.json',
 'sample_data']

In [ ]:
files.download('columns.json')